### ISSUE TITLE
Terraform created internet network endpoint group does not work with load balancer

### ISSUE THREAD

**Issue by villesau** [hashicorp/terraform-provider-google#11657]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to the `modular-magician` user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to `hashibot`, a community member has claimed the issue already.

### Terraform Version

1.0.11

### Affected Resource(s)

* google_compute_global_network_endpoint_group

### Terraform Configuration Files

https://serverfault.forumming.com/question/958/google-cloud-external-load-balancer-39-s-backend-services-randomly-fail-with-502-server-error

### Debug Output

-

### Panic Output

-
### Expected Behavior

The setup should work

### Actual Behavior

LB returns 502

### Steps to Reproduce

Run following config:
https://serverfault.forumming.com/question/958/google-cloud-external-load-balancer-39-s-backend-services-randomly-fail-with-502-server-error

### Important Factoids

- creating identical setup from the gke UI works fine
- importing it to terraform without changes works
- creating 1:1 similar setup via terraform directly succeeds, but the setup does not work. More info in the reference links.

### References

- https://serverfault.forumming.com/question/958/google-cloud-external-load-balancer-39-s-backend-services-randomly-fail-with-502-server-error
- https://stackoverflow.com/questions/72146395/load-balancer-giving-failed-to-pick-backend-with-internet-network-group

**Comment by slevenick** [comment]
It looks like there is an answer that has been accepted for those issues that includes adding a `depends_on` block, does that work for you?

**Comment by villesau** [comment]
Yes that worked. But figuring something like that out is pure luck. Most of the google results pointed out to missing health checks which are not required by this setup. I think there should be more clear dependencies between the resources. For example, maybe `google_compute_global_network_endpoint` could be defined inside `google_compute_global_network_endpoint_group` rather than separately? Or some other way to sort it out.

**Comment by pawelJas** [comment]
@slevenick @villesau This issue is quite old and it seems the problem was addressed by @slevenick. Can we close this?

**Comment by jbbqqf** [comment]
Hi! Triaging older NEG issues — this one looks ready to close.

The original reporter confirmed in [the thread](https://github.com/hashicorp/terraform-provider-google/issues/11657#issuecomment-1121530177) that adding an explicit `depends_on` between `google_compute_backend_service` and the `google_compute_global_network_endpoint` resources resolves the 502s, and another community member [pinged for closure earlier this year](https://github.com/hashicorp/terraform-provider-google/issues/11657#issuecomment-3710784439).

This is the well-known Terraform attribute-vs-resource-dependency pattern — when a backend service references the *group* but the actual *endpoint* hasn't been written yet, the LB sees an empty backend. The provider can't infer the dependency because the endpoint side-loads into the group through a separate resource. Working snippet:

```hcl
resource "google_compute_backend_service" "lb_backend" {
  name     = "lb-backend"
  protocol = "HTTPS"

  backend {
    group = google_compute_global_network_endpoint_group.neg.id
  }

  depends_on = [google_compute_global_network_endpoint.endpoint]
}
```

Provider docs: https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/compute_global_network_endpoint

Would you mind closing this out? If you've hit a different flavour of the bug where `depends_on` doesn't help, open a fresh issue with a current repro and I'll dig.

(Disclosure: I drafted this comment with help from Claude Code while triaging stale issues; the resolution and pattern were verified manually against the issue thread and provider docs.)

**Comment by villesau** [comment]
If it still implicitly relies on `depends_on`, I think it's not ready to be closed since it is way too unclear.
